File: chronosynd-py/chronosynd_py/features/syscall_ngram.py
"""Disjoint-window 1-gram syscall histogram extractor, parity-graded
mirror of `chronosynd-rs/crates/features/src/syscall_ngram.rs`. Both
sides assert byte-identical feature vectors on the shared test vectors"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass

import numpy as np

from chronosynd_py.core import (
    FloatArray,
    InvalidParameterError,
)


@dataclass(frozen=True, slots=True)
class EmittedFeatures:
    """A feature vector emitted when a per-process window closes.
    `process_key` identifies the process and `feature_vector` is the dense
    numeric representation the scoring engine consumes"""

    process_key: str
    feature_vector: list[float]


class SyscallNgramExtractor:
    """Disjoint-window 1-gram histogram extractor over a fixed syscall
    vocabulary. The emitted feature vector has dimension `len(vocab) + 1`,
    with the trailing slot counting out-of-vocab syscalls"""

    def __init__(self, vocab: Sequence[int], window_size: int) -> None:
        """Build with an explicit syscall vocabulary and disjoint window
        size. Duplicates in `vocab` raise so the feature contract stays
        unambiguous"""
        if window_size < 1:
            raise InvalidParameterError(
                f"window_size must be at least 1, got {window_size}"
            )
        vocab_index: dict[int, int] = {}
        for idx, syscall_nr in enumerate(vocab):
            if syscall_nr in vocab_index:
                raise InvalidParameterError(
                    f"duplicate vocabulary entry {syscall_nr} at index {idx}"
                )
            vocab_index[syscall_nr] = idx
        self._vocab_index = vocab_index
        self._feature_dim = len(vocab) + 1
        self._window_size = window_size
        self._per_process: dict[str, list[float]] = {}
        self._pending: dict[str, int] = {}

    @property
    def feature_dim(self) -> int:
        """Dimension of every emitted feature vector, constant across calls"""
        return self._feature_dim

    @property
    def other_bucket_index(self) -> int:
        """Index in the feature vector reserved for syscalls outside vocab"""
        return self._feature_dim - 1

    @property
    def tracked_process_count(self) -> int:
        """Number of processes the extractor is currently accumulating for"""
        return len(self._per_process)

    def accumulate(
        self, process_key: str, syscall_nr: int
    ) -> EmittedFeatures | None:
        """Feed one observed syscall. Returns an `EmittedFeatures` when the
        per-process window closes, `None` while the window is still open"""
        counts = self._per_process.get(process_key)
        if counts is None:
            counts = [0.0] * self._feature_dim
            self._per_process[process_key] = counts

        bucket = self._vocab_index.get(syscall_nr, self._feature_dim - 1)
        counts[bucket] += 1.0

        pending = self._pending.get(process_key, 0) + 1
        self._pending[process_key] = pending

        if pending < self._window_size:
            return None

        denom = float(self._window_size)
        normalized = [count / denom for count in counts]
        emitted = EmittedFeatures(
            process_key=process_key, feature_vector=normalized
        )

        self._per_process[process_key] = [0.0] * self._feature_dim
        self._pending[process_key] = 0
        return emitted

    def reset_process(self, process_key: str) -> None:
        """Discard any in-progress window for a process. The next event for
        that key starts a fresh window"""
        self._per_process.pop(process_key, None)
        self._pending.pop(process_key, None)


def feature_vector_to_array(features: EmittedFeatures) -> FloatArray:
    """Convert an `EmittedFeatures` into a 1-D numpy array. Kept here so
    callers that mix the harness with the extractor do not have to know
    the storage shape `feature_vector` uses internally"""
    return np.asarray(features.feature_vector, dtype=np.float64)

File: chronosynd-rs/crates/features/src/syscall_ngram.rs
//! Syscall 1-gram histogram extractor over per-process disjoint windows,
//! every `window_size` events the extractor emits a normalized histogram
//! sized `vocab.len() + 1` (the trailing slot is the out-of-vocab bucket)

use std::collections::HashMap;

use crate::error::{FeatureError, Result};
use crate::types::EmittedFeatures;

/// Disjoint-window 1-gram histogram extractor over a fixed syscall vocabulary
#[derive(Debug)]
pub struct SyscallNgramExtractor {
    vocab_index: HashMap<u32, usize>,
    feature_dim: usize,
    window_size: usize,
    per_process: HashMap<String, Vec<f64>>,
    pending: HashMap<String, usize>,
}

impl SyscallNgramExtractor {
    /// Build with an explicit syscall vocabulary and disjoint window size,
    /// the emitted feature vector has dimension `vocab.len() + 1` with the
    /// trailing slot counting out-of-vocab syscalls
    pub fn new(vocab: Vec<u32>, window_size: usize) -> Result<Self> {
        if window_size == 0 {
            return Err(FeatureError::InvalidParameter(
                "window_size must be at least 1".into(),
            ));
        }
        let mut vocab_index = HashMap::with_capacity(vocab.len());
        for (idx, syscall_nr) in vocab.iter().enumerate() {
            if vocab_index.insert(*syscall_nr, idx).is_some() {
                return Err(FeatureError::DuplicateVocabEntry(*syscall_nr));
            }
        }
        let feature_dim = vocab.len() + 1;
        Ok(Self {
            vocab_index,
            feature_dim,
            window_size,
            per_process: HashMap::new(),
            pending: HashMap::new(),
        })
    }

    /// Dimension of every emitted feature vector, stays constant for the
    /// lifetime of the extractor
    pub fn feature_dim(&self) -> usize {
        self.feature_dim
    }

    /// Index in the feature vector reserved for syscalls outside the vocabulary
    pub fn other_bucket_index(&self) -> usize {
        self.feature_dim - 1
    }

    /// Number of processes the extractor is currently tracking
    pub fn tracked_process_count(&self) -> usize {
        self.per_process.len()
    }

    /// Feed one observed syscall, returns `Some(EmittedFeatures)` when the
    /// per-process window closes, otherwise `None`
    pub fn accumulate(&mut self, process_key: &str, syscall_nr: u32) -> Option<EmittedFeatures> {
        let counts = self
            .per_process
            .entry(process_key.to_string())
            .or_insert_with(|| vec![0.0; self.feature_dim]);
        let bucket = self
            .vocab_index
            .get(&syscall_nr)
            .copied()
            .unwrap_or(self.feature_dim - 1);
        counts[bucket] += 1.0;

        let pending = self
            .pending
            .entry(process_key.to_string())
            .or_insert(0);
        *pending += 1;

        if *pending < self.window_size {
            return None;
        }

        let denom = self.window_size as f64;
        let normalized: Vec<f64> = counts.iter().map(|count| count / denom).collect();
        let emitted = EmittedFeatures {
            process_key: process_key.to_string(),
            feature_vector: normalized,
        };

        self.per_process.insert(process_key.to_string(), vec![0.0; self.feature_dim]);
        self.pending.insert(process_key.to_string(), 0);
        Some(emitted)
    }

    /// Discard any in-progress window for a process, the next event for
    /// that process starts a fresh window
    pub fn reset_process(&mut self, process_key: &str) {
        self.per_process.remove(process_key);
        self.pending.remove(process_key);
    }
}
